bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Bookie.java

```diff
@@ -441,117 +335,187 @@ private void checkIfDirsOnSameDiskPartition(List<File> dirs) throws DiskPartitio
         }
     }
 
+    static Versioned<Cookie> readAndVerifyCookieFromRegistrationManager(Cookie masterCookie,
+                                                                        ServerConfiguration conf,
+                                                                        RegistrationManager rm)
+            throws BookieException {
+        // we need to loop through all possible bookie identifiers to ensure it is treated as a new environment
+        // just because of bad configuration
+        List<BookieSocketAddress> addresses = Lists.newArrayListWithExpectedSize(3);
+        try {
+            // ip address
+            addresses.add(getBookieAddress(
+                new ServerConfiguration(conf).setUseHostNameAsBookieID(false).setAdvertisedAddress(null)));
+            // host name
+            addresses.add(getBookieAddress(
+                new ServerConfiguration(conf).setUseHostNameAsBookieID(true).setAdvertisedAddress(null)));
+            // advertised address
+            if (null != conf.getAdvertisedAddress()) {
+                addresses.add(getBookieAddress(conf));
+            }
+        } catch (UnknownHostException e) {
+            throw new UnknownBookieIdException(e);
+        }
+        Versioned<Cookie> rmCookie = null;
+        for (BookieSocketAddress address : addresses) {
+            try {
+                rmCookie = Cookie.readFromRegistrationManager(rm, address);
+                // If allowStorageExpansion option is set, we should
+                // make sure that the new set of ledger/index dirs
+                // is a super set of the old; else, we fail the cookie check
+                if (conf.getAllowStorageExpansion()) {
+                    masterCookie.verifyIsSuperSet(rmCookie.getValue());
+                } else {
+                    masterCookie.verify(rmCookie.getValue());
+                }
+            } catch (CookieNotFoundException e) {
+                continue;
+            }
+        }
+        return rmCookie;
+    }
+
+    private static Pair<List<File>, List<Cookie>> verifyAndGetMissingDirs(Cookie masterCookie,
```

Comment:
Is it valid to have a bookie with more than one cookie? Shouldn't all cookies match the value read from registration manager?

Reply:
in the `allowStorageExpansion`, there might be at a point that there might be two cookies (e.g. a disk failure on storage expansion, some directory is not updated yet).

Reply:
This would imply that a bookie started one time with a unavailable directory while had previously been available, which sounds like an error case. The bookie shouldn't have come up with a disk missing.

Reply:
what is 'a unavailable directory'? this method is only to verify existing cookies that available in local dirs and collecting directories that miss cookies. I am not sure we are on the same page about this.

Reply:
What I'm saying is that there shouldn't be a case where one directory was down, the rest had their cookie upgraded, and then that directory came back. If a directory is down, it should will throw a FileNotFoundException, and prevent boot. So it shouldn't be possible to have different cookies on a single system in most cases.

The only case I can think of where more than one cookie exists is where a failure occurred in the process of writing the cookies. In this case, we should require manual intervention.

Reply:
```
What I'm saying is that there shouldn't be a case where one directory was down, the rest had their cookie upgraded, and then that directory came back. If a directory is down, it should will throw a FileNotFoundException, and prevent boot. So it shouldn't be possible to have different cookies on a single system in most cases.
```
the case you mentioned here won't happen. if a directory is missed, it is caught with directory missed cookie. If a directory missed cookie is in an old environment, that validation will fail. 

```
The only case I can think of where more than one cookie exists is where a failure occurred in the process of writing the cookies. In this case, we should require manual intervention.
```
Thas was what I explained. We don't need manual intervention (and it makes things become complicated). A restart of the bookie should continue with storage expansion and succeed eventually. IMO, we should not change any operational behavior at this pull request because it only tends to address the root cause first.

Reply:
> Thas was what I explained. We don't need manual intervention (and it makes things become complicated). A restart of the bookie should continue with storage expansion and succeed eventually.

This won't be expansion though, because if the cookie exists, then the directory is non-empty so can't be used for expansion.

Reply:
if a new directory was not written with a cookie during storage expansion. the storage expansion will continue to fix it eventually. if an existing dir's cookie is not updated during storage expansion, there are two cases : 1) a bookie starts again without expansion, that the verification will fail 2) a bookie starts again with expansion, it is okay to continue as well because the master cookie and rm cookie validation is still valid (because of the superset). that's still a healthy environment.